PPC64LE Alpine Edge ICE

Starting today I have gotten a consistent internal compiler error while trying to build b2. Is it possible to update the compiler (this one seems to be about a year old patched version) in the container? I'm not sure what changed to cause this. Alpine's docker was last updated 12 days ago, and B2 has not been updated since last release. Below is an example log:

```
Run source ci/github/install.sh
  source ci/github/install.sh
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    GIT_FETCH_JOBS: 8
    NET_RETRY_COUNT: 5
    B2_CI_VERSION: 1
    B2_VARIANT: release
    B2_LINK: static
    LCOV_BRANCH_COVERAGE: 0
    CODECOV_NAME: Github Actions
    ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION: true
    B2_USE_CCACHE: 0
    DEBIAN_FRONTEND: noninteractive
    GHA_USE_NODE_20: true
    BDDE_ARCH: ppc64le
    BDDE_DISTRO: alpine
    BDDE_EDITION: edge
    B2_WRAPPER: bdde
    B2_DONT_EMBED_MANIFEST: 1
    B2_ADDRESS_MODEL: 
    B2_COMPILER: gcc
    B2_CXXSTD: 14
    B2_SANITIZE: 
    B2_STDLIB: 
    B2_DEFINES: BOOST_DECIMAL_QEMU_TEST
++ BOOST_CI_TARGET_BRANCH=refs/heads/develop
++ export BOOST_CI_TARGET_BRANCH=develop
++ BOOST_CI_TARGET_BRANCH=develop
++ export BOOST_CI_SRC_FOLDER=/home/runner/work/decimal/decimal
++ BOOST_CI_SRC_FOLDER=/home/runner/work/decimal/decimal
++ echo BOOST_CI_TARGET_BRANCH=develop
++ echo BOOST_CI_SRC_FOLDER=/home/runner/work/decimal/decimal
++ [[ '' == \y\e\s ]]
+++ dirname ci/github/install.sh
++ . ci/github/../common_install.sh
+++ set -ex
+++++ dirname ci/github/../common_install.sh
++++ cd ci/github/..
++++ pwd
+++ CI_DIR=/home/runner/work/decimal/decimal/ci
+++ '[' 0 == 1 ']'
+++ print_on_gha '::group::Setup B2 variables'
+++ . /home/runner/work/decimal/decimal/ci/enforce.sh
Setup B2 variables
Checkout and setup Boost build tree
+++ print_on_gha '::group::Setup B2'
Setup B2
+++ [[ '' != \1 ]]
+++ print_on_gha '::group::Bootstrap B2'
Bootstrap B2
  +++ trap show_bootstrap_log ERR
  +++ '[' '!' -f b2 ']'
  +++ bdde ./bootstrap.sh
  ++ id -u
  ++ id -g
  + docker run --rm --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -v /home/runner/work/decimal/boost-root:/boost:rw -v /home/runner/work/decimal/decimal/bdde:/bdde:ro -v /home/runner/.vimrc:/home/boost/.vimrc:ro -v /etc/passwd:/etc/passwd:ro -v /etc/group:/etc/group:ro -v /etc/shadow:/etc/shadow:ro --user 1001:128 --workdir /boost/. -i -e B2_ADDRESS_MODEL -e B2_CI_VERSION -e B2_COMPILER -e B2_CXXSTD -e B2_DEFINES -e B2_DONT_EMBED_MANIFEST -e B2_JOBS -e B2_LINK -e B2_SANITIZE -e B2_STDLIB -e B2_TOOLSET -e B2_USE_CCACHE -e B2_VARIANT -e B2_WRAPPER docker.io/jeking3/bdde:alpine-edge-ppc64le-latest /bin/bash -c ./bootstrap.sh
  Unable to find image 'jeking3/bdde:alpine-edge-ppc64le-latest' locally
  alpine-edge-ppc64le-latest: Pulling from jeking3/bdde
  26011e7cd874: Pulling fs layer
  a1d7eb56ea88: Pulling fs layer
  bd1f3c804b3d: Pulling fs layer
  9689f7c7f3ab: Pulling fs layer
  1dbd4f7ae472: Pulling fs layer
  4fbd9d139d37: Pulling fs layer
  7b51cce2d4fd: Pulling fs layer
  ae52c2cae1e9: Pulling fs layer
  99f341511b52: Pulling fs layer
  a3b8678f3c9c: Pulling fs layer
  a5f72ff305cc: Pulling fs layer
  1ff0e3721e76: Pulling fs layer
  ac63c9bb46e7: Pulling fs layer
  f6eae27973c0: Pulling fs layer
  561947d310d9: Pulling fs layer
  1f58c727c35f: Pulling fs layer
  2d29fa3ae96e: Pulling fs layer
  8a9a8e0804f1: Pulling fs layer
  dd8235ab0d36: Pulling fs layer
  5fdf8835ac71: Pulling fs layer
  7b51cce2d4fd: Waiting
  ae52c2cae1e9: Waiting
  99f341511b52: Waiting
  a3b8678f3c9c: Waiting
  a5f72ff305cc: Waiting
  1ff0e3721e76: Waiting
  ac63c9bb46e7: Waiting
  f6eae27973c0: Waiting
  561947d310d9: Waiting
  1f58c727c35f: Waiting
  2d29fa3ae96e: Waiting
  9689f7c7f3ab: Waiting
  1dbd4f7ae472: Waiting
  4fbd9d139d37: Waiting
  5fdf8835ac71: Waiting
  8a9a8e0804f1: Waiting
  dd8235ab0d36: Waiting
  a1d7eb56ea88: Verifying Checksum
  a1d7eb56ea88: Download complete
  26011e7cd874: Download complete
  1dbd4f7ae472: Verifying Checksum
  1dbd4f7ae472: Download complete
  9689f7c7f3ab: Verifying Checksum
  9689f7c7f3ab: Download complete
  26011e7cd874: Pull complete
  4fbd9d139d37: Verifying Checksum
  4fbd9d139d37: Download complete
  7b51cce2d4fd: Verifying Checksum
  7b51cce2d4fd: Download complete
  ae52c2cae1e9: Verifying Checksum
  ae52c2cae1e9: Download complete
  99f341511b52: Verifying Checksum
  99f341511b52: Download complete
  a3b8678f3c9c: Verifying Checksum
  a3b8678f3c9c: Download complete
  a5f72ff305cc: Verifying Checksum
  a5f72ff305cc: Download complete
  ac63c9bb46e7: Verifying Checksum
  ac63c9bb46e7: Download complete
  a1d7eb56ea88: Pull complete
  1ff0e3721e76: Verifying Checksum
  1ff0e3721e76: Download complete
  561947d310d9: Verifying Checksum
  561947d310d9: Download complete
  f6eae27973c0: Verifying Checksum
  f6eae27973c0: Download complete
  1f58c727c35f: Verifying Checksum
  1f58c727c35f: Download complete
  8a9a8e0804f1: Verifying Checksum
  8a9a8e0804f1: Download complete
  dd8235ab0d36: Download complete
  2d29fa3ae96e: Verifying Checksum
  2d29fa3ae96e: Download complete
  5fdf8835ac71: Verifying Checksum
  5fdf8835ac71: Download complete
  bd1f3c804b3d: Verifying Checksum
  bd1f3c804b3d: Download complete
  bd1f3c804b3d: Pull complete
  9689f7c7f3ab: Pull complete
  1dbd4f7ae472: Pull complete
  4fbd9d139d37: Pull complete
  7b51cce2d4fd: Pull complete
  ae52c2cae1e9: Pull complete
  99f341511b52: Pull complete
  a3b8678f3c9c: Pull complete
  a5f72ff305cc: Pull complete
  1ff0e3721e76: Pull complete
  ac63c9bb46e7: Pull complete
  f6eae27973c0: Pull complete
  561947d310d9: Pull complete
  1f58c727c35f: Pull complete
  2d29fa3ae96e: Pull complete
  8a9a8e0804f1: Pull complete
  dd8235ab0d36: Pull complete
  5fdf8835ac71: Pull complete
  Digest: sha256:397fc1f8f0244c8a8bf90b661d3fc949fd4406d698921a5e5d0cb4f2971a481b
  Status: Downloaded newer image for jeking3/bdde:alpine-edge-ppc64le-latest
  Building B2 engine..
  
  ###
  ###
  ### Using 'gcc' toolset.
  ###
  ###
  
  g++ (Alpine 13.2.1_git20240309) 13.2.1 20240309
  Copyright (C) 2023 Free Software Foundation, Inc.
  This is free software; see the source for copying conditions.  There is NO
  warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  
  
  ###
  ###
  
  > g++ -x c++ -std=c++11 -pthread -O2 -s -DNDEBUG bindjam.cpp builtins.cpp class.cpp command.cpp compile.cpp constants.cpp cwd.cpp debug.cpp debugger.cpp events.cpp execcmd.cpp execnt.cpp execunix.cpp filent.cpp filesys.cpp fileunix.cpp frames.cpp function.cpp glob.cpp hash.cpp hcache.cpp hdrmacro.cpp headers.cpp jam_strings.cpp jam.cpp jamgram.cpp lists.cpp make.cpp make1.cpp md5.cpp mem.cpp modules.cpp native.cpp option.cpp output.cpp parse.cpp pathnt.cpp pathsys.cpp pathunix.cpp regexp.cpp rules.cpp scan.cpp search.cpp startup.cpp tasks.cpp timestamp.cpp value.cpp variable.cpp w32_getreg.cpp mod_command_db.cpp mod_db.cpp mod_jam_builtin.cpp mod_jam_class.cpp mod_jam_errors.cpp mod_jam_modules.cpp mod_order.cpp mod_path.cpp mod_property_set.cpp mod_regex.cpp mod_sequence.cpp mod_set.cpp mod_string.cpp mod_summary.cpp mod_sysinfo.cpp mod_version.cpp -o b2
  g++: internal compiler error: Segmentation fault signal terminated program cc1plus
  Please submit a full bug report, with preprocessed source (by using -freport-bug).
  See <https://gitlab.alpinelinux.org/alpine/aports/-/issues> for instructions.
  
  Failed to build B2 build engine
  ++++ show_bootstrap_log
  ++++ cat bootstrap.log
  cat: bootstrap.log: No such file or directory
  Error: Process completed with exit code 1.
```

Complete CI run is here: https://github.com/cppalliance/decimal/actions/runs/13550829714/job/37873772787?pr=883#step:13:1

--- a/README.md
+++ b/README.md
@@ -29,7 +29,7 @@ you must satisfy the prerequisites of running a
 [multiarch](https://github.com/multiarch/qemu-user-static) docker container:
 
 1. Install the binfmt-support and qemu-user-static packages.
-2. Run `docker run --rm --privileged multiarch/qemu-user-static --reset -p yes` or run `bdde-multiarch`.
+2. Run `bdde-multiarch`.
 
 Due to the performance decrease when running multiarch emulation, the CI script does not attempt
 to run asan, tsan, ubsan, or build all of boost on every platform.  The GitHub workflow file defines
@@ -200,7 +200,7 @@ on a little-endian x86_64 host running Ubuntu Bionic:
 ### Installing Prerequisites
 
     user@ubuntu:~$ sudo apt-get install -y binfmt-support qemu-user-static
-    user@ubuntu:~$ sudo docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+    user@ubuntu:~$ bdde-multiarch
 
 ### Running a unit test in Boost.Predef while emulating a big-endian system
 

--- a/bin/linux/bdde-build
+++ b/bin/linux/bdde-build
@@ -18,5 +18,7 @@ if [ ! -z "${BDDE_REBUILD}" ]; then
   DOCKER_CACHE_OPTION=--no-cache
 fi
 
+${BDDE_ROOT}/bdde-multiarch --version
+
 set -x
 docker build ${DOCKER_CACHE_OPTION} -t ${BDDE_IMAGE} -f Dockerfile.${BDDE_DISTRO}-${BDDE_EDITION}-${BDDE_ARCH} .

--- a/bin/linux/bdde-multiarch
+++ b/bin/linux/bdde-multiarch
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 #
-# Copyright (C) 2018 - 2024 James E. King III
+# Copyright (C) 2018 - 2025 James E. King III
 #
 # Use, modification, and distribution are subject to the
 # Boost Software License, Version 1.0. (See accompanying file
@@ -10,5 +10,18 @@
 # Prepare the docker environment to run multiarch images.
 #
 
-set -x
-docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+# NOTE: See https://github.com/multiarch/qemu-user-static/issues/212 for why we use tonistiigi/binfmt instead.
+
+#BDDE_MULTIARCH="multiarch/qemu-user-static --reset -p yes"
+BDDE_MULTIARCH=tonistiigi/binfmt:qemu-v9.2.2-52
+
+if [ "$1" == "--version" ]; then
+  docker run --privileged --rm ${BDDE_MULTIARCH} --version
+else
+  set -x
+  # changed from multiarch/qemu-user-static to tonistiigi/binfmt
+  # because the multiarch docker image hasn't been updated
+  # see https://github.com/multiarch/qemu-user-static/issues/212
+  docker run --privileged --rm ${BDDE_MULTIARCH} --uninstall "qemu-*"
+  docker run --rm --privileged ${BDDE_MULTIARCH} --install all
+fi

